<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Import-Sample-Data" data-toc-modified-id="Import-Sample-Data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Import Sample Data</a></span><ul class="toc-item"><li><span><a href="#Housing-Price" data-toc-modified-id="Housing-Price-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Housing Price</a></span></li><li><span><a href="#Tax-Data" data-toc-modified-id="Tax-Data-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Tax Data</a></span></li></ul></li><li><span><a href="#join-Data-Frame-Method" data-toc-modified-id="join-Data-Frame-Method-2"><span class="toc-item-num">2&nbsp;&nbsp;</span><code>join</code> Data Frame Method</a></span></li><li><span><a href="#merge-Data-Frame-Method" data-toc-modified-id="merge-Data-Frame-Method-3"><span class="toc-item-num">3&nbsp;&nbsp;</span><code>merge</code> Data Frame Method</a></span></li><li><span><a href="#concat-Pandas-Function" data-toc-modified-id="concat-Pandas-Function-4"><span class="toc-item-num">4&nbsp;&nbsp;</span><code>concat</code> Pandas Function</a></span></li></ul></div>

In [71]:
import pandas as pd

## Import Sample Data

### Housing Price
The Housing Price data by zip from Zillow was downloaded from https://www.zillow.com/research/data/ .  


In [72]:
# house_df = pd.read_csv('../data/Zip_zhvi_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv')
house_df = pd.read_csv('https://raw.githubusercontent.com/emsanqui/Python_Training/master/data/Zip_zhvi_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv')
house_df.tail()

RegionID  SizeRank  RegionName RegionType StateName State         City   
27010     78282     39809       46799        zip        IN    IN   Zanesville  \
27011     80861     39809       52163        zip        IA    IA     Protivin   
27012     69074     39809       26576        zip        WV    WV   Farmington   
27013     64515     39809       16245        zip        PA    PA    Oak Ridge   
27014     80190     39809       50160        zip        IA    IA  Martensdale   

                                Metro        CountyName  2000-01-31  ...   
27010                    Bluffton, IN      Wells County         NaN  ...  \
27011                             NaN     Howard County         NaN  ...   
27012                    Fairmont, WV     Marion County         NaN  ...   
27013                  Pittsburgh, PA  Armstrong County         NaN  ...   
27014  Des Moines-West Des Moines, IA     Warren County         NaN  ...   

          2022-06-30     2022-07-31     2022-08-31     2022-09-30   
27010  172931.113553  173726.818013  173951.262842  173561.093030  \
27011  100986.105306  102843.677816  104486.379416  105810.880094   
27012   98430.653830   99565.281599  100539.187287  100590.832913   
27013   79001.623917   79797.115905   80327.757426   80096.167777   
27014  182921.063195  184139.412398  184688.403061  184729.848799   

          2022-10-31     2022-11-30     2022-12-31     2023-01-31   
27010  173267.350240  173316.329830  173324.855024  173343.968822  \
27011  106454.239900  106897.177835  107196.786820  107037.013712   
27012  100489.571763  100123.791299  100196.541945  100556.335407   
27013   79256.847170   78066.159714   77466.901397   77711.217346   
27014  185179.263239  185869.441655  186269.605436  186267.600047   

          2023-02-28     2023-03-31  
27010  173739.502719  174765.029798  
27011  106366.668071  105680.453415  
27012  101320.684556  102093.120933  
27013   77581.056657   76642.895385  
27014  186353.463783  187071.766317  

[5 rows x 288 columns]

Since the housing data has 288 columns, we will select a few columns to work with.

In [73]:
house_df.columns

Index(['RegionID', 'SizeRank', 'RegionName', 'RegionType', 'StateName',
       'State', 'City', 'Metro', 'CountyName', '2000-01-31',
       ...
       '2022-06-30', '2022-07-31', '2022-08-31', '2022-09-30', '2022-10-31',
       '2022-11-30', '2022-12-31', '2023-01-31', '2023-02-28', '2023-03-31'],
      dtype='object', length=288)

Below we renamed the `RegionID` columns to `ZIPCODE` and select the `ZIPCODE`, `2023-01-31`, `2023-02-28`, and `2023-03-31` columns.  Then, we use the `groupby()` and `mean()` functions to calculate the means grouped by `ZIPCODE`.

Note that the `groupby()` function will indexed the results by the columns passed to the function.  In this case `ZIPCODE`.

In [74]:
house_df = (house_df
            .rename(columns={'RegionID':'ZIPCODE'})
            [['ZIPCODE','2023-01-31','2023-02-28','2023-03-31']]
            .groupby('ZIPCODE')
            .mean()
           )
house_df

2023-01-31     2023-02-28     2023-03-31
ZIPCODE                                             
58196    289183.650560  290142.126012  291579.076221
58197    452812.730506  453702.600673  456075.368414
58200    349605.853667  349729.762203  350587.034195
58201    402072.180437  401473.914859  401888.086022
58202    318573.721730  317472.656484  317132.928828
...                ...            ...            ...
777366   155659.981192  157189.842481  159339.001500
787970   525168.720819  524857.684731  527158.578055
787971   576787.298849  572369.045592  570487.717820
787972   418133.934425  418081.008836  419930.686376
808739   467690.950571  470487.848802  476631.332425

[27015 rows x 3 columns]

### Tax Data 
The tax data by zip code from the IRS was downloaded from https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2020-zip-code-data-soi

Below we import the data and assign the `ZIPCODE` column as the index.

In [ ]:
# tax_df = pd.read_csv('../data/20zpallnoagi.csv')
tax_df = pd.read_csv('https://raw.githubusercontent.com/emsanqui/Python_Training/master/data/20zpallnoagi.csv')
tax_df.tail()

Since the tax data has 164 columns, we will select a few columns to work with.

In [ ]:
tax_df.columns

Below we select the `ZIPCODE`, `A11902`, `N12000`, and `A12000` columns.  Then, we use the `groupby()` and `mean()` functions to calculate the means grouped by `ZIPCODE`.

In [ ]:
tax_df = (tax_df[['ZIPCODE','A11902','N12000','A12000']]
          .groupby('ZIPCODE')
          .mean()
         )
tax_df

## `join` Data Frame Method

Below we `inner` join the housing and tax data frames on `ZIPCODE`.  Other types of join types include `left`, `right`, `outer`, `inner`, `cross`, the default is `left`.  This works like a sql query join operator.

More information is available from the documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.join.html

In [ ]:
house_df.join(tax_df, on='ZIPCODE', how='inner').tail()

## `merge` Data Frame Method

The `merge` operator is more flexible than `join`.  However, `merge` requiring more detailed inputs.

Below we use the `merge` method to perform the same operation as above with `join`.

More information is available from the documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html

In [ ]:
house_df.reset_index().merge(right=tax_df.reset_index(), how='inner', on='ZIPCODE').tail().set_index('ZIPCODE')

## `concat` Pandas Function

More information is available from the documentation https://pandas.pydata.org/docs/reference/api/pandas.concat.html

The `concat` function can concatenate two data frames as below.

In [ ]:
df1 = pd.DataFrame([['a', 1], ['b', 2]], columns=['letter', 'number'])
print('This is the `df1` data frame')
display(df1)

df2 = pd.DataFrame([['c', 3], ['d', 4]], columns=['letter', 'number'])
print('This is the `df2` data frame')
display(df2)

print('Below we concatenate the `df2` data frame to `df1` data frame')
pd.concat([df1, df2])

By default the `concat()` function will concatenate rows (`axis=0`).  However, we can also concatenate columns by setting `axis=1`.

In [ ]:
pd.concat([df1, df2], axis=1)

We can also concatenate a series to a data frame as below.

In [ ]:
df3 = pd.DataFrame({'a': 1, 'b': 2}, index=[0])
print('This is the `df3` data frame')
display(df7)

new_row = pd.Series({'a': 3, 'b': 4})
print('This is the `new_row` series')
display(new_row)

print('Below we concatenated `new_row` to `df3`')
pd.concat([df3, new_row.to_frame().T], ignore_index=True)